src/utils.py

```diff
@@ -66,41 +69,61 @@ def get_machine_token(contract_token: str, contracts_url=DEFAULT_CONTRACTS_URL,
         "Authorization": f"Bearer {contract_token}",
         "Content-Type": "application/json",
     }
-    try:
-        data = make_request(
-            "POST",
-            f"{contracts_url}/v1/context/machines/token",
-            data=json.dumps(payload),
-            headers=headers,
-            timeout=60,
-        )
-        return data.get("machineToken", "")
-    except requests.exceptions.RequestException:
-        return None
-    except KeyError:
-        return None
-
 
-def get_resource_token(machine_token, contracts_url=DEFAULT_CONTRACTS_URL, proxies=None):
+    with tempfile.NamedTemporaryFile(prefix="ca", suffix="cert", delete=False) as ca_tempfile:
+        ca_filename = None
+        if ca_certificate is not None:
+            ca_tempfile.write(ca_certificate)
+            ca_tempfile.close()
+            ca_filename = ca_tempfile.name
+        try:
+            data = make_request(
+                "POST",
+                f"{contracts_url}/v1/context/machines/token",
+                data=json.dumps(payload),
+                headers=headers,
+                timeout=60,
+                verify=ca_filename,
+            )
+            return data.get("machineToken", "")
+        except requests.exceptions.RequestException:
+            return None
+        except KeyError:
+            return None
+        finally:
+            os.unlink(ca_tempfile.name)
```

Comment:
even with delete=False?

Reply:
Seems it's a bit [more complicated](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile). In any case, why set delete=False? We unlink it right after anyway, let the context manager handle that.

Reply:
I think it's because of the Python version. If you check the docs v3.10 (which is the same version that the charm is tested on, I think), there's only a `delete` option which defaults to `True` and that means the file will be deleted as soon as it's closed. Since, we need to close the file inside the managed context (because the `requests` library need to be able to *read* the file), we have to set `delete=False` and manually delete the final at the end.